src/sqlfluff/core/linter/linted_file.py

```diff
@@ -0,0 +1,491 @@
+"""Defines the LintedFile class.
+
+This holds linting results for a single file, and also
+contains all of the routines to apply fixes to that file
+post linting.
+"""
+
+import os
+import logging
+from typing import (
+    Any,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from benchit import BenchIt
+
+from sqlfluff.core.errors import (
+    SQLBaseError,
+    SQLLintError,
+    CheckTuple,
+)
+from sqlfluff.core.string_helpers import findall
+from sqlfluff.core.templaters import TemplatedFile
+
+# Classes needed only for type checking
+from sqlfluff.core.parser.segments.base import BaseSegment, FixPatch
+
+from sqlfluff.core.linter.common import NoQaDirective, EnrichedFixPatch
+
+# Instantiate the linter logger
+linter_logger: logging.Logger = logging.getLogger("sqlfluff.linter")
+
+
+class LintedFile(NamedTuple):
+    """A class to store the idea of a linted file."""
+
+    path: str
+    violations: List[SQLBaseError]
+    time_dict: dict
+    tree: Optional[BaseSegment]
+    ignore_mask: List[NoQaDirective]
+    templated_file: TemplatedFile
+
+    def check_tuples(self) -> List[CheckTuple]:
+        """Make a list of check_tuples.
+
+        This assumes that all the violations found are
+        linting violations (and therefore implement `check_tuple()`).
+        If they don't then this function raises that error.
+        """
+        vs: List[CheckTuple] = []
+        v: SQLLintError
+        for v in self.get_violations():
+            if hasattr(v, "check_tuple"):
+                vs.append(v.check_tuple())
+            else:
+                raise v
+        return vs
+
+    def get_violations(
+        self,
+        rules: Optional[Union[str, Tuple[str, ...]]] = None,
+        types: Optional[Union[Any, Iterable[Any]]] = None,
+        filter_ignore: bool = True,
+        fixable: bool = None,
+    ) -> list:
+        """Get a list of violations, respecting filters and ignore options.
+
+        Optionally now with filters.
+        """
+        violations = self.violations
+        # Filter types
+        if types:
+            try:
+                types = tuple(types)
+            except TypeError:

```

Comment:
What input are we expecting that would cause a `TypeError` as we are handling here?

Reply:
Great question, I think it's singular violation types. That code path is covered by tests, so it's being used, but I'll see if I can either refactor this - or at least add an explanatory comment